First cut at the revocation section (#49)
AnonCreds-WG/anoncreds-spec: spec/data_flow_revocation.md

### AnonCreds Revocation Data Flow

AnonCreds includes a mechanism that supports the revocation of verifiable
credentials. This mechanism includes:

- An [[ref: issuer]] setting up to issue revocable credentials.
- An [[ref: issuer]] issuing revocable credentials.
- An [[ref: issuer]] activating or revoking issued credentials.
- A [[ref: verifier]] requesting a presentation to include a non-revocation proof
  for one or more revocable credentials.

> - A [[ref: holder]] generating based on the request of the verifier a

Reviewer: ```suggestion
- A [[ref: holder]] generating based on the presentation request of the verifier a
```

Author: Fixed.

  non-revocation proof for attributes derived from revocable credentials.
- A [[ref: verifier]] verifying a non-revocation proof included in a
  presentation from a [[ref: holder]].

A fundamental goal of AnonCreds is to not provide a correlatable identifier for

> a either a [[ref: holder]] or a credential as part of generation and verification of an AnonCreds

Reviewer: ```suggestion
either a [[ref: holder]] or a credential as part of generation and verification of an AnonCreds
```

Author: Fixed.

presentation. Applying that goal to revocation means that the revocation
mechanism must support the [[ref: holder]] proving a credential used in generating a
presentation is not revoked without providing a correlatable identifier for that

> credential. As such, the AnonCreds revocation mechanism uses a Zero Knowledge

Reviewer: ```suggestion
credential or the holder itself. As such, the AnonCreds revocation mechanism uses a Zero Knowledge
```

Author: Fixed.

Proof (ZKP) that allows the [[ref: holder]] to prove a credential they hold is not

> revoked without revealing an identifier for their credential.

Reviewer: ```suggestion
revoked without revealing an identifier for their credential or the holder.
```

#### AnonCreds Issuer Setup With Revocation

The details of [[ref: issuer]] setting up revokable credential types are covered
in the [issuer
setup](#issuer-create-and-publish-revocation-registry-objects) section
of this specification.

#### AnonCreds Issuance with Revocation

The details of an [[ref: issuer]] issuing a revokable credential to a [[ref:
holder]] are covered in the [issuance data
flow](data_flow_issuance.md#issue-credential) section of this specification.

> If the [[ref: issuer]] has created the [[ref: RevReg]] from which they are
> issuing credentials to have an `ISSUANCE_TYPE` of `ISSUE_ON_DEMAND`, the
> [[ref:issuer]] must publish a [[ref: RevRegEntry]] (as described in the [next
> section](#anoncreds-credential-activationrevocation-and-publication)) as each
> credential is issued. Based on the experience of the AnonCreds community in the
> use of revocable credentials, it is highly recommended the `ISSUE_ON_DEMAND`
> approach **NOT** be used unless absolutely required by your use case. The reason
> for this recommendation is that `ISSUE_ON_DEMAND` requires near real-time
> activation, meaning the issuing of each credential requires the
> near-simultaneous publication of a new [[ref: RevRegEntry]], resulting in many
> [[ref: RevRegEntry]] transactions being performed, one per credential issued.
> 
> Further, if a credential contains some kind of "Issue Date" attribute in the
> credential, verifiers could use that attribute information combined with the
> publication of a [[ref: RevRegEntry]] at the same time to learn the ID of the
> holder's credential, giving them both a correlatable identifier for the holder
> and a way to monitor if the holder's credential is revoked.
> 
> For these reasons we anticipate the deprecation or removal of the
> `ISSUE_ON_DEMAND` approach in the next version of AnonCreds specification.
> Feedback from the community on this would be appreciated. We are particularly
> interested in understanding what use cases there are for `ISSUE_ON_DEMAND`.

Reviewer: ```suggestion
::: warning
If the [[ref: issuer]] has created the [[ref: RevReg]] from which they are
issuing credentials to have an `ISSUANCE_TYPE` of `ISSUE_ON_DEMAND`, the
[[ref:issuer]] must publish a [[ref: RevRegEntry]] (as described in the [next
section](#anoncreds-credential-activationrevocation-and-publication)) as each
credential is issued. Based on the experience of the AnonCreds community in the
use of revocable credentials, it is highly recommended the `ISSUE_ON_DEMAND`
approach **NOT** be used unless absolutely required by your use case. The reason
for this recommendation is that `ISSUE_ON_DEMAND` requires near real-time
activation, meaning the issuing of each credential requires the
near-simultaneous publication of a new [[ref: RevRegEntry]], resulting in many
[[ref: RevRegEntry]] transactions being performed, one per credential issued.

Further, if a credential contains some kind of "Issue Date" attribute in the
credential, verifiers could use that attribute information combined with the
publication of a [[ref: RevRegEntry]] at the same time to learn the ID of the
holder's credential, giving them both a correlatable identifier for the holder
and a way to monitor if the holder's credential is revoked.

For these reasons we anticipate the deprecation or removal of the
`ISSUE_ON_DEMAND` approach in the next version of AnonCreds specification.
Feedback from the community on this would be appreciated. We are particularly
interested in understanding what use cases there are for `ISSUE_ON_DEMAND`.
:::
```

Reviewer: Added a "warning" highlight box, since we do not recommend the usage of `ISSUE_ON_DEMAND`. I am also wondering, if we should not move this entire paragraph to the credential definition setup part with revocation capabilities. 

We than have to distinguish the optional revocation part within each section into a) `ISSUANCE_PER_DEFAULT` and b) `ISSUE_ON_DEMAND`. 

Or we only mention the two issuing modes once at the credential definition setup part and emphasize, that *ISSUE_ON_DEMAND' probably will not be supported in future versions and therefore don`t explain the details. 

Not sure yet what the best approach here is...

Author: I like the idea of mentioning it with the warning in the setup section and have moved it there.  I think I will leave a warning and pointer to the primary warning.

Author: Adjusted.  I moved the warning into the data setup, gave the warning a header, and then linked to that section in a couple of other places in the spec. I think that works pretty well.

#### AnonCreds Credential Activation/Revocation and Publication

When an [[ref: issuer]] decides to revoke a previously issued credential (or
activate a previously inactive/revoked credential), they do so by publishing
another instance of the [[ref: RevRegEntry]] object. Recall from the issuer
setup section, the specification about [creating and publishing the first [[ref:
RevRegEntry]]](data_flow_setup.md#creating-the-initial-revocation-registry-entry-object)
for a [[ref: RevReg]]. In that process, the accumulator for the initial state of
the [[ref: RevReg]] is published. When subsequent [[ref: RevRegEntry]]
transactions are published to the ledger, each includes an updated value of the

> accumulator, along with lists of the identifiers of the credentials that have
> had their revocation state toggled -- either changed to issued, or changed to
> revoked -- since the last [[ref: RevRegEntry]] was published. An example of the
> data in the [[ref: RevRegEntry]] is shown in the following example of a [[ref:
> RevRegEntry]], pulled from [this transaction on the Sovrin
> MainNet](https://indyscan.io/tx/SOVRIN_MAINNET/domain/140326).

Reviewer: ```suggestion
accumulator. The update of the accumulator is necessary with each revocation or activation of a credential or set of credentials since the last published [[ref: RevRegEntry]]. This is because only the factors (all factors are listed in the respective tails file) of credentials which are active (meaning not being revoked) contribute to the accumulator. Therefore in addition to the updated accumulator value, every RevRegEntry contains a list of indexes of credential factors which have been revoked within the [[ref: RevRegEntry]]. This list of factor indexes is a so called [[ref: Witness Delta]] and enables the [[ref: Holder]] to successfully generate a proof of non revocation.

An example of the
data in the [[ref: RevRegEntry]] is shown in the following example of a [[ref:
RevRegEntry]], pulled from [this transaction on the Sovrin
MainNet](https://indyscan.io/tx/SOVRIN_MAINNET/domain/140326).
```

Reviewer: I tried to differentiate between "identifiers of the credentials" and the "index of the factor within a tails file". From my point of view, this is crucial, since imho it is not some kind of ID of a credential, that is published within a RevRegEntry, but rather the index of the factor, that is bound to the credential. 

But I am not entirely happy with my wording here...

Author: I consider the "RevocationID" to be the combined key `index+RevRegId`, and that the RevocationId is essentially a `CredentialId`.

**Question**: Does the `Witness Delta` include the factor contributed by the Credential that is the subject of the Non-Revocation Proof?  I've always wondered about that.  I think of it as the `Accumulator` = `Witness Delta` * `Credential Factor`

Author: I liked your wording and put it in verbatim. I think we can merge with that and adjust if needed.

Reviewer: > I consider the "RevocationID" to be the combined key `index+RevRegId`, and that the RevocationId is essentially a `CredentialId`.
> 
> **Question**: Does the `Witness Delta` include the factor contributed by the Credential that is the subject of the Non-Revocation Proof? I've always wondered about that. I think of it as the `Accumulator` = `Witness Delta` * `Credential Factor`

_This is handled by requiring accumulator updates to also publish a witness delta as part of the same transaction. This delta tells provers how to adjust their witness (referencing other indexes in the public tails file) to bring it back into harmony with the current value of the accumulator, such that the updated witness times the private factor once again equals the accumulator value._

(got it from [here](https://hyperledger-indy.readthedocs.io/projects/hipe/en/latest/text/0011-cred-revocation/README.html))

Author: OK, so the witness is the product of all of the unrevoked factors excluding the one for the credential held by the holder. But the witness delta is exactly as you used it in your note above.  I've added the Witness Delta to the terminology section.

In looking at that reference, I also documented how the NRP ZKP is generated.  I hope it is enough detail, although I've noted that more detail is needed on generating the commitments.

  `not revoked`) since the last [[ref: RevRegEntry]] was published to the ledger.
- `revoked`: an array (possibly empty or not supplied) of the indices within the
  [[ref: RevReg]] of the credentials whose state has changed to `revoked` (also known as
  `not active`) since the last [[ref: RevRegEntry]] was published to the ledger.

In the example transaction above no credentials are `issued` (meaning changed
from status `revoked` to `issued`) and only one, the credential with index
`172`, is changed to `revoked`. Both lists can have an arbitrary number of
entries, up to the total number of credentials in the [[ref: RevReg]].

The algorithm to calculate the value of a [[ref: RevRegEntry]] accumulator at
any time is the same: determine the (modulo) product of the primes for each
non-revoked credential in the [[ref: REV_REG]], as described
[here](#creating-the-initial-revocation-registry-entry-object).

A [[ref: VDR]] publishing the [[ref: RevRegEntry]] transactions MAY perform its own calculation
of the accumulator to ensure that the calculation of the accumulator after all
of the revocation status updates to the credentials within the [[ref: RevReg]] have been

> applied, rejecting the transaction if the calculated accumulator does not match that from the [[ref: issuer]].

Reviewer: Sounds wise to me, that the VDR does a double check on the updated accumulator value, but do we want to mention it here (in the AnonCreds Spec) or is it out of scope?

If the [[ref: issuer]] and [[ref: VDR]] get out of sync about the status of
credentials in the [[ref: RevReg]], such that the two cannot agree on the current value
of the accumulator, the [[ref:issuer]] must rationalize the differences and
produce a [[ref: RevRegEntry]] transaction that accounts for both the last published
[[ref: RevRegEntry]] published in the [[ref: VDR]] and the desired revocation status of all of the
credentials in the [[ref: RevReg]].

Note that the [[ref: holder]] is not involved in the credential revocation
process. There is no technical requirement for an [[ref: issuer]] to notify the
[[ref: holder]] that a credential they were issued has been revoked. That said, it is a
courtesy that may improve the user experience of the [[ref: holder]]. [Aries RFC 0183
Revocation
Notification](https://github.com/hyperledger/aries-rfcs/tree/main/features/0183-revocation-notification)
is an example of how that can be done. Even if not notified by the [[ref:
issuer]] of the revocation of a credential, the [[ref: holder]] can detect their
credential has been revoked when they retrieve the list of revoked credentials

> from the [[ref: VDR]] and discover the index of their credential in the list.

Reviewer: Sounds wise to me, that the VDR does a double check on the updated accumulator value, but do we want to mention it here (in the AnonCreds Spec) or is it out of scope?

I think the same about getting the VDR and issuer revocation list syncing and about informing the user about that his credential has been revoked. On the las one we are rather on the Aries level, than the Indy / AnonCreds level imho?

Author: I think it is in scope as if the VDR (e.g. where ever it is published) and the issuer are not in sync, the holder will not be able to create a valid NRP. 

OK -- updated to reflect:

- If things get out of sync the NRP can't be created by the holder
- the VDR may prevent going out of sync
- the issuer must fix things if they do go out of sync.

I think that is reasonable to put into the spec.

Reviewer: Sounds reasonable to me. I think I would add a `::: note` wrapper around it, just to indicate, that is some kind of "additional" info the reader gets here. Or maybe even a `::: warning` if we consider it as very important. 

To me the fact that the issuer has to keep things in sync is crucial because that is in the heart of AnonCreds. The info, that the VDR can assist not so much, because that depends on where (which VDR) AnonCreds are implemented in / to. How the issuer can fix when being out of sync (in Indy) sounds to me like an "problems / fixing" section. But maybe that is just me...

(Just trying to keep the AnonCreds part as much reduced as possible while adding "additional" separately) 😄

Author: Good comments.  I've put a note wrapper on that as you suggest.

Creating an AnonCreds presentation is a two-step process, beginning with a
request from the [[ref: verifier]] asking the [[ref: holder]] to include a
non-revocation proof (NRP) in the presentation, and then the [[ref: holder]]
creating the NRP and including it in the presentation sent to the [[ref:
verifier]]. Both steps are covered here.

Both the verifier requesting a non-revocation proof, and the [[ref: holder]]
generating the non-revocation proof are covered in the sections of this
specification about [requesting](#request-non-revocation-proofs) and
[generating](#generate-non-revocation-proofs) presentations, respectively.

#### AnonCreds Verification with Revocation

A [[ref: verifier]] receives the presentation from the [[ref: holder]] and
processes the [non-revocation-related parts of the presentation](#verify-presentation) and
the [revocation-related parts of the presentation](#verify-non-revocation-proof)
(if any) in the presentation. The resulting status of the presentation combines the
verification outcomes from processing all proofs within the presentation. If 
verification of one or more of the embedded proofs is unsuccessful, the

> presentation is rejected as unverifiable.

Reviewer: Personally, I would move these sections to the "Create Presentation Request" and "Verify Presentation" sections.

Author: I moved the meat of the revocation sections over to the presentation and verification docs. I updated the wording to make it clear that this overview is only there to say -- "go look somewhere else".  I think that is reasonable.

presentation. Applying that goal to revocation means that the revocation
mechanism must support the [[ref: holder]] proving a credential used in generating a
presentation is not revoked without providing a correlatable identifier for that
credential. As such, the AnonCreds revocation mechanism uses a Zero Knowledge
Proof (ZKP) that allows the [[ref: holder]] to prove a credential they hold is not
revoked without revealing an identifier for their credential.

#### AnonCreds Issuer Setup With Revocation

The details of [[ref: issuer]] setting up revokable credential types are covered
in the [issuer
setup](data_flow_setup.md#generating-a-creddef-with-revocation-support) section
of this specification.

#### AnonCreds Issuance with Revocation

The details of an [[ref: issuer]] issuing a revokable credential to a [[ref:
holder]] are covered in the [issuance data
flow](data_flow_issuance.md#issue-credential) section of this specification.

If the [[ref: issuer]] has created the [[ref: RevReg]] from which they are
issuing credentials to have an `ISSUANCE_TYPE` of `ISSUE_ON_DEMAND`, the
[[ref:issuer]] must publish a [[ref: RevRegEntry]] (as described in the next

> section) as each credential issued. Based on the experience of the AnonCreds

Author: Instead changed to "as each credential is issued."

creating the NRP and including it in the presentation sent to the [[ref:
verifier]]. Both steps are covered here.

**NOTE**: Often in discussions about verifiable presentations, the term "[[ref: prover]]"
is used to indicate the participant generating the request presentation.
Throughout the Hyperledger Indy AnonCreds implementation the term `[[ref: prover]]` is
used in the names of methods performed out by that participant. However, because
in AnonCreds the [[ref: holder]] and the [[ref: prover]] are always the same entity, we'll use
[[ref: holder]] to refer to the participant generating the requested presentation to
emphasize that the same entity is both issued credentials and generating the
presentations from those credentials.

##### Verifier Revocation Interval Request

The inclusion of a "Revocation Interval" by the [[ref: verifier]] in a presentation
request is a minimal extension to the Presentation Request data format that is
defined in the [Create Presentation Request](#create-presentation-request)
section of this document. Notably the `non_revoked` data item is added:

```json
"non_revoked" : {
    "from" : <Unix Epoch time>
    "to" : <Unix Epoch time>
}
```

> The item may at the outer level of the presentation request, such that it

Reviewer 2: ```suggestion
The item may be at the outer level of the presentation request, such that it
```

Reviewer 2: sense check paragraph

Author: Reviewed and clarified, with missing word added.

  credentials in the registry is `active` or `revoked`. This information is part
  of the [[ref: RevReg]], and so likely cached by the [[ref: holder]] when they were first issued
  the credential.
- The tails file for [[ref: RevReg]], the location (a URL) of which is stored in the
  issued credential's [[ref: RevReg]]. The [[ref: holder]] likely (though not necessarily) would
  have collected the tails file at the time of issuance. Recall (from this
  section of the specification) that the tails file for a [[ref: RevReg]] is generated at
  creation time and does not change.
- The index of the credential within the [[ref: RevReg]] for the [[ref: holder]]'s specific
  credential being used in the presentation. This information is given to the
  [[ref: holder]] by the [[ref: issuer]] when the verifiable credential is issued.
- The accumulator published by the [[ref: issuer]] for the [[ref: RevRegEntry]] that the [[ref: holder]]
  will use in generating the NRP. In the Hyperledger Indy implementation of
  AnonCreds, the entries are published on the ledger, and collected via a
  special request to the ledger (detailed below).

> - The revocation status changes of all of the credential indices up to the

Reviewer 2: delete `of`? Not sure. Sense check para

:::

- `x_list`" is ...
  - `rho`" is ...
  - `r`" is ...
  - `r_prime`" is ...
  - `r_prime_prime`" is ...
  - `r_prime_prime_prime`" is ...
  - `o`" is ...
  - `o_prime`" is ...
  - `m`" is ...
  - `m_prime`" is ...
  - `t`" is ...
  - `t_prime`" is ...
  - `m2`" is ...
  - `s`" is ...
  - `c`" is ...
- `c_list`" is ...
  - `e`" is ...
  - `d`" is ...
  - `a`" is ...
  - `g`" is ...
  - `w`" is ...
  - `s`" is ...
  - `u`" is ...

As well, in the presentation data model, added to the `identifiers` item, is the
timestamp (Unix epoch format) of the [[ref: RevRegEntry]] used to construct the NRP
(see example below). The [[ref: verifier]] needs the `rev_reg_id` and `timestamp` to get
the correct accumulator to use in verifying the NRP.

```json
"identifiers": [
    {
        "schema_id": "7BPMqYgYLQni258J8JPS8K:2:degree schema:46.58.87",
        "cred_def_id": "7BPMqYgYLQni258J8JPS8K:3:CL:70:faber.agent.degree_schema",
        "rev_reg_id": "7BPMqYgYLQni258J8JPS8K:4:7BPMqYgYLQni258J8JPS8K:3:CL:70:faber.agent.degree_schema:CL_ACCUM:61d5a381-30be-4120-9307-b150b49c203c",
        "timestamp": 1656269796
    }
]
```

> As always, the [[ref: holder]]'s presentation, with the embedded NRP(s), is set to the

Author: fixed.

each.

The [[ref: verifier]] begins by extracting from the section of the presentation
for a given revocable credential the `non_revoc_proof` and `identifiers` data
items. The [[ref: verifier]] must retrieve (possibly from its cache, otherwise
from the [[ref: VDR]]) the published [[ref: RevRegEntry]] given the `rev_reg_id`
and `timestamp` values from the `identifiers` data item. The [[ref: verifier]]
extracts the `accumulator` item from the [[ref: RevRegEntry]] retrieved. Note
that the [[ref: verifier]] does not need to collect the revocation status of all
of the credentials in the registry, nor the contents of the tails file for the
[[ref: RevReg]]. Only the issuer and [[ref: holder]] needs that data. During the
verification process, the [[ref: verifier]] does not learn the index of the
[[ref: holder]]'s credential in the [[ref: RevReg]].

Once the [[ref: verifier]] gets the data in the `non_revoc_proof` data item from
the presentation for the NRP being processed, plus the accumulator from
appropriate [[ref: RevRegEntry]], the following steps are carried out to verify
the NRP.

:::todo
To Do: Outline the NRP verification calculation.
:::

:::todo

> To Do: Is there a separate process to bind the NRP to the credential?

Author: Not sure -- I think this needs to be an issue that is investigated. I'll add it.

:::

The verification code MUST surface to the [[ref: verifier]] if any part of the

> presentation, including any PoNR(s), fail cryptographic verification. The

Author: fixed.

> presentation, including any PoNR(s), fail cryptographic verification. The

Author: FYI - I'm making the fix in my copy vs. accepting the commit because I'm not sure how sign off is done. I don't want to screw up the DCO.  Someday I'll try it... :-)

verification code MAY surface additional detail about what part of the
presentation failed, such as which NRP failed verification (if any).

The [[ref: verifier]] SHOULD evaluate the presentation to make sure that the
[[ref: holder]] provided all requested NRPs. Notably, if any expected PoNRs
are not received in the presentation, the [[ref: verifier]] SHOULD check to see
if the given credential type is revocable. If not, it is acceptable that no
NRP was received. However, if the credential used in the generation of the

> proof is revocable, and the [[ref: holder]] did not provide the P0N-R, the

Reviewer 2: ```suggestion
proof is revocable, and the [[ref: holder]] did not provide the NRP, the
```

Author: Fixed.
